### PR TITLE
ARROW-5402: [Plasma] add support to pin objects in plasma store

### PR DESCRIPTION
**Pull request by zhijunfu**
This adds support to pin an object in plasma store, which is useful in some scenarios (e.g. we want to ensure the actor creation task arguments won't be evicted from plasma so that actor can be re-constructed when it crashes).  

This is done by specifying a flag when creating an object, if the flag is set, then the object won't be managed by LRU cache in plasma, and thus won't be evicted even when its reference count drops to zero, it can only be removed by an explicit `Delete()` call.

@pcmoritz @robertnishihara @guoyuhong  could you please help take a look? thanks.

### CODE REVIEW

File: cpp/src/plasma/common.h [hunk 1]
```diff
@@ -107,7 +107,8 @@ struct ObjectTableEntry {
   int64_t create_time;
   /// How long creation of this object took.
   int64_t construct_duration;
-
+  /// Whether this object is included in LRU cache.
```
Comment (guoyuhong) [inline cpp/src/plasma/common.h]:
```suggestion
  /// Whether this object is excluded in LRU cache.
```

**Review by guoyuhong — COMMENTED** [review]
There is no a separate API to pin the object , so the pinned property could only be set at the creation time, right?

**Comment by zhijunfu** [comment]
> There is no a separate API to pin the object , so the pinned property could only be set at the creation time, right?

Yes, this allows us to ensure the object cannot be evicted right after its creation, if it's implemented as a separate API there's still a window that the object can be evicted. We can add another separate API later if we have such scenarios that specifying the flag at creation doesn't meet the requirement.

File: cpp/src/plasma/client.cc [hunk 2]
```diff
@@ -208,10 +208,11 @@ class PlasmaClient::Impl : public std::enable_shared_from_this<PlasmaClient::Imp
                  int num_retries = -1);
 
   Status Create(const ObjectID& object_id, int64_t data_size, const uint8_t* metadata,
-                int64_t metadata_size, std::shared_ptr<Buffer>* data, int device_num = 0);
+                int64_t metadata_size, std::shared_ptr<Buffer>* data, int device_num = 0,
+                bool is_pinned = true);
```
[the remark this answers is not in the document]
Comment (zhijunfu) [reply cpp/src/plasma/client.cc]: Good catch, thanks!  Yes these should be aligned, as you mentioned `is_pinned = false` is intended.  The flag was previously call `use_lru` which default to `true`, when converting to `is_pinned` which default to `false`, some of the places are overlooked...

**Comment by guoyuhong** [comment]
Looks good to me now. How do you think of this PR, @pcmoritz @robertnishihara ?

**Comment by emkornfield** [comment]
@pcmoritz @robertnishihara what are the next steps on this?

**Comment by emkornfield** [comment]
@pcmoritz @robertnishihara ping?  @zhijunfu it looks like this needs a rebase?

**Comment by emkornfield** [comment]
Closing as no response from my previous ping.

**Pull request closed without merging**